File: Macro/PS0-3/ps0.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import fredapi
import os
from datetime import datetime
import seaborn as sns

# Set up paths
root_path = "~/Dropbox/Schoolwork/Macro1/ProblemSets/PS0"
code_path = os.path.join(root_path, "code")
data_path = os.path.join(root_path, "Data")
output_path = os.path.join(root_path, "Output")

# Initialize FRED API
fred = fredapi.Fred(api_key='3937d3e1a400536de448b9415c9114d9')

# Set plot style
plt.style.use('seaborn')

# Problem 2
def problem_2():
    # Part A
    pcend = fred.get_series('PCEND')
    pces = fred.get_series('PCES')
    gdp = fred.get_series('GDP')
    
    # Filter data after 2005
    start_date = '2005-01-01'
    df = pd.DataFrame({
        'PCEND': pcend,
        'PCES': pces,
        'GDP': gdp
    }).loc[start_date:]
    
    df['PCENDS'] = df['PCEND'] + df['PCES']
    df['pce_percent_gdp'] = df['PCENDS'] / df['GDP']
    
    plt.figure(figsize=(10, 6))
    plt.plot(df.index, df['pce_percent_gdp'])
    plt.title('Consumption of Non-Durables and Services: % of GDP')
    plt.xlabel('Time')
    plt.ylabel('Percentage of GDP')
    plt.savefig(os.path.join(output_path, 'graph2a.png'))
    plt.close()
    
    # Parts B through G follow similar pattern...

# Problem 3
def problem_3():
    # Part A
    gdp_pc = fred.get_series('A939RX0Q048SBEA')
    df = pd.DataFrame({'GDP': gdp_pc})
    df['rGDP_pc'] = np.log(df['GDP'])
    
    # Create date numeric values for regression
    df['date_num'] = (df.index - df.index[0]).days
    
    # Fit linear regression
    from sklearn.linear_model import LinearRegression
    X = df['date_num'].values.reshape(-1, 1)
    y = df['rGDP_pc'].values
    model = LinearRegression()
    model.fit(X, y)
    df['trend'] = model.predict(X)
    df['dt_GDP'] = df['rGDP_pc'] - df['trend']
    
    # Plot results
    plt.figure(figsize=(10, 6))
    plt.plot(df.index, df['rGDP_pc'])
    plt.plot(df.index, df['trend'])
    plt.title('Log Real GDP per Capita with Trend')
    plt.savefig(os.path.join(output_path, '3aa.png'))
    plt.close()
    
    # Parts B through E follow similar pattern...

# Problem 4-8 follow similar patterns...

if __name__ == "__main__":
    problem_2()
    # problem_3()
    # etc...

File: Macro/PS0-3/ps0.r
# Required libraries to be installed first
library(tidyverse)  # for data manipulation and plotting
library(fredr)      # for FRED API access
library(lubridate)  # for date handling
library(readxl)     # for Excel file reading
library(zoo)        # for time series operations

# Set working directory paths
root_path <- "~/Library/CloudStorage/Dropbox/Schoolwork/Macro1/ProblemSets/PS0"
code_path <- file.path(root_path, "code")
data_path <- file.path(root_path, "Data")
output_path <- file.path(root_path, "Output")

# Set FRED API key
fredr_set_key("3937d3e1a400536de448b9415c9114d9")

# Problem 2
if (TRUE) {
  # Part A
  # Get individual series
  pcend <- fredr_series_observations("PCEND")
  pces <- fredr_series_observations("PCES")
  gdp <- fredr_series_observations("GDP")
  
  # Combine the series
  df <- pcend %>%
    select(date, value) %>%
    rename(PCEND = value) %>%
    left_join(
      pces %>% select(date, value) %>% rename(PCES = value),
      by = "date"
    ) %>%
    left_join(
      gdp %>% select(date, value) %>% rename(GDP = value),
      by = "date"
    )
  
  # Filter and calculate
  df <- df %>%
    filter(date >= as.Date("2005-01-01")) %>%
    mutate(
      PCENDS = PCEND + PCES,
      pce_percent_gdp = PCENDS / GDP
    )
  
  # Create plot
  ggplot(df, aes(x = date, y = pce_percent_gdp)) +
    geom_line() +
    labs(
      x = "Time",
      y = "Percentage of GDP",
      title = "Consumption of Non-Durables and Services: % of GDP"
    ) +
    theme_minimal()
  ggsave(file.path(output_path, "graph2a.png"))
  
  # Part B
  pcedg <- fredr_series_observations("PCEDG")
  gpdi <- fredr_series_observations("GPDI")
  
  df_b <- pcedg %>%
    select(date, value) %>%
    rename(PCEDG = value) %>%
    left_join(
      gpdi %>% select(date, value) %>% rename(GPDI = value),
      by = "date"
    ) %>%
    left_join(
      gdp %>% select(date, value) %>% rename(GDP = value),
      by = "date"
    )
  
  df_b <- df_b %>%
    filter(date >= as.Date("2005-01-01")) %>%
    mutate(
      PCEDGI = PCEDG + GPDI,
      durinv_percent_gdp = PCEDGI / GDP
    )
  
  ggplot(df_b, aes(x = date, y = durinv_percent_gdp)) +
    geom_line() +
    labs(
      x = "Time",
      y = "Percentage of GDP",
      title = "Consumption of Durables and Investment: % of GDP"
    ) +
    theme_minimal()
  ggsave(file.path(output_path, "graph2b.png"))
  
  # Similar pattern for parts C through G...
}

# Problem 3
if (FALSE) {
  # Part A
  gdp_pc <- fredr_series_observations("A939RX0Q048SBEA")
  
  df <- gdp_pc %>%
    mutate(
      rGDP_pc = log(value),
      date_num = as.numeric(date)
    )
  
  # Fit linear model
  model <- lm(rGDP_pc ~ date_num, data = df)
  df$dt_GDP <- residuals(model)
  
  # Create plots
  p1 <- ggplot(df, aes(x = date, y = rGDP_pc)) +
    geom_line() +
    geom_smooth(method = "lm", se = FALSE) +
    labs(
      x = "Time",
      y = "Log Real GDP per Capita",
      title = "Real GDP per Capita with Trend"
    ) +
    theme_minimal()
  ggsave(file.path(output_path, "3aa.png"), p1)
  
  p2 <- ggplot(df, aes(x = date, y = dt_GDP)) +
    geom_line() +
    labs(
      x = "Time",
      y = "Detrended Log Real GDP per Capita",
      title = "Detrended Real GDP per Capita"
    ) +
    theme_minimal()
  ggsave(file.path(output_path, "3ab.png"), p2)
  
  # Parts B through E follow similar pattern...
}
